File: src/main/java/eu/happybe/openapi/OpenAPI.java
package eu.happybe.openapi;

import eu.happybe.openapi.api.bossbar.BossBarBuilder;
import eu.happybe.openapi.api.form.FormQueue;
import eu.happybe.openapi.mysql.DatabaseData;
import eu.happybe.openapi.mysql.QueryQueue;
import eu.happybe.openapi.mysql.query.LazyRegisterQuery;
import eu.happybe.openapi.party.PartyManager;
import eu.happybe.openapi.ranks.RankDatabase;
import eu.happybe.openapi.scoreboard.packets.RemoveObjectivePacket;
import eu.happybe.openapi.scoreboard.packets.SetDisplayObjectivePacket;
import eu.happybe.openapi.scoreboard.packets.SetScorePacket;
import eu.happybe.openapi.servers.ServerManager;
import eu.happybe.openapi.utils.PlayerUtils;
import io.gomint.config.YamlConfig;
import io.gomint.event.EventListener;
import io.gomint.plugin.Plugin;
import lombok.Getter;
import lombok.SneakyThrows;

public class OpenAPI extends Plugin implements EventListener {

    @Getter
    private static OpenAPI instance;

    @SneakyThrows
    @Override
    public void onStartup() {
        OpenAPI.instance = this;
        // TODO - Save config

        this.registerListener(this);

        YamlConfig config = new YamlConfig(this.dataFolder() + "/config.yml").load().saveToMap();
        DatabaseData.update();
    }

    public void onEnable() {
        OpenAPI.instance = this;
        this.saveResource("/config.yml");

        this.getServer().getPluginManager().registerEvents(this, this);

        DatabaseData.update(
            this.getConfig().getString("mysql-host"),
            this.getConfig().getString("mysql-user"),
            this.getConfig().getString("mysql-password")
        );

        RankDatabase.init();
        ServerManager.init();

        this.getServer().getNetwork().registerPacket((byte)0x6a, RemoveObjectivePacket.class);
        this.getServer().getNetwork().registerPacket((byte)0x6b, SetDisplayObjectivePacket.class);
        this.getServer().getNetwork().registerPacket((byte)0x6c, SetScorePacket.class);
    }

    @Override
    public void onDisable() {
        ServerManager.save();
    }

    @EventHandler
    public void onLogin(PlayerLoginEvent event) {
        Player player = event.getPlayer();

        QueryQueue.submitQuery(new LazyRegisterQuery(event.getPlayer().getName()), (query -> {
            RankDatabase.savePlayerRank(player, String.valueOf(((LazyRegisterQuery)query).row.get("Rank")));
            PartyManager.handleLoginQuery(player, (LazyRegisterQuery) query);
            PlayerUtils.updateNameTag(player);
        }));
    }

    @EventHandler
    public void onFormHandle(DataPacketReceiveEvent event) {
        if(event.getPacket() instanceof ModalFormResponsePacket) {
            if(FormQueue.handleFormDataPacket(event.getPlayer(), (ModalFormResponsePacket) event.getPacket())) {
                event.setCancelled();
            }
        }
    }

    @EventHandler
    public void onQuit(PlayerQuitEvent event) {
        BossBarBuilder.removeBossBar(event.getPlayer());
        FormQueue.handleQuit(event.getPlayer());
        PartyManager.handleQuit(event.getPlayer());
    }
}

File: src/main/java/eu/happybe/openapi/OpenAPI.kt
package eu.happybe.openapi

import cn.nukkit.event.EventHandler
import cn.nukkit.event.Listener
import cn.nukkit.event.player.PlayerLoginEvent
import cn.nukkit.event.player.PlayerQuitEvent
import cn.nukkit.event.server.DataPacketReceiveEvent
import cn.nukkit.network.protocol.ModalFormResponsePacket
import cn.nukkit.plugin.PluginBase
import eu.happybe.openapi.bossbar.BossBarBuilder
import eu.happybe.openapi.event.LoginQueryReceiveEvent
import eu.happybe.openapi.form.FormQueue
import eu.happybe.openapi.mysql.AsyncQuery
import eu.happybe.openapi.mysql.DatabaseData
import eu.happybe.openapi.mysql.QueryQueue
import eu.happybe.openapi.mysql.query.LazyRegisterQuery
import eu.happybe.openapi.party.PartyManager
import eu.happybe.openapi.ranks.RankDatabase
import eu.happybe.openapi.scoreboard.packets.RemoveObjectivePacket
import eu.happybe.openapi.scoreboard.packets.SetDisplayObjectivePacket
import eu.happybe.openapi.scoreboard.packets.SetScorePacket
import eu.happybe.openapi.servers.ServerManager
import eu.happybe.openapi.utils.PlayerUtils
import lombok.Getter

class OpenAPI : PluginBase(), Listener {
    override fun onEnable() {
        instance = this
        this.saveResource("/config.yml")
        server.pluginManager.registerEvents(this, this)
        DatabaseData.update(
                config.getString("mysql-host"),
                config.getString("mysql-user"),
                config.getString("mysql-password")
        )
        RankDatabase.init()
        ServerManager.init()
        server.network.registerPacket(0x6a.toByte(), RemoveObjectivePacket::class.java)
        server.network.registerPacket(0x6b.toByte(), SetDisplayObjectivePacket::class.java)
        server.network.registerPacket(0x6c.toByte(), SetScorePacket::class.java)
    }

    override fun onDisable() {
        ServerManager.save()
    }

    @EventHandler
    fun onLogin(event: PlayerLoginEvent) {
        val player = event.player
        QueryQueue.submitQuery(LazyRegisterQuery(event.player.name)) { query: AsyncQuery ->
            RankDatabase.savePlayerRank(player, (query as LazyRegisterQuery).row["Rank"].toString())
            PartyManager.handleLoginQuery(player, query)
            PlayerUtils.updateNameTag(player)
            server.pluginManager.callEvent(LoginQueryReceiveEvent(player, query))
        }
    }

    @EventHandler
    fun onFormHandle(event: DataPacketReceiveEvent) {
        if (event.packet is ModalFormResponsePacket) {
            if (FormQueue.handleFormDataPacket(event.player, event.packet as ModalFormResponsePacket)) {
                event.setCancelled()
            }
        }
    }

    @EventHandler
    fun onQuit(event: PlayerQuitEvent) {
        BossBarBuilder.removeBossBar(event.player)
        FormQueue.handleQuit(event.player)
        PartyManager.handleQuit(event.player)
    }

    companion object {
        @Getter
        private var instance: OpenAPI? = null
    }
}